Reimplement context menu

D0m1nos opened the pull request:
This pull request aims to reimplement the context menu removed after the new ui.

Right-clicking on a song in the song list lets you add it to a playlist (will be implemented in #60 ) or put it as the next song in the queue (if the queue exists).

Right-clicking on a song in the queue lets you remove it or add it to a playlist.

![image](https://github.com/user-attachments/assets/cf5b235e-0fe2-48a4-bd7b-95790390d3e3)

D0m1nos commented:
I just noticed that if the context menu is too far to the right or too far down it will get clipped. Will try to fix tomorrow.

![image](https://github.com/user-attachments/assets/10f60b47-6a08-4de5-aa27-45bcdbe88719)

Tnixc commented on src/renderer/src/components/song/context-menu/items/PlayNext.tsx:
```diff
@@ -1,28 +1,23 @@
+      }}
+    >
+      <p>Play next</p>
+      <i class="ri-menu-add-line"></i>
```
Please use icons from 'lucide-solid' instead.  Remix icons are to be phased out in #83

D0m1nos replied:
Done. Replaced with `<ListStartIcon />`

Tnixc commented on src/renderer/src/components/song/song-list/SongList.tsx:
```diff
@@ -88,15 +95,32 @@ const SongList: Component<SongViewProps> = (props) => {
+          }}
+        >
+          <p>Add to playlist</p>
+          <i class="ri-add-line"></i>
```
Same here

D0m1nos replied:
Done. Replaced with `<PlusIcon />`

Tnixc commented on src/renderer/src/components/song/song-queue/SongQueue.tsx:
```diff
@@ -105,14 +108,31 @@ const SongQueue: Component = () => {
+          }}
+        >
+          <p>Remove from queue</p>
+          <i class="ri-delete-back-2-line"></i>
```
And here.

D0m1nos replied:
Done. Replaced with `<PlusIcon />` and `<DeleteIcon />`

brw commented:
shouldn't this also use the existing Tooltip component? That should also fix this as that's handled by Floating UI
> I just noticed that if the context menu is too far to the right or too far down it will get clipped. Will try to fix tomorrow.

D0m1nos commented:
> shouldn't this also use the existing Popover component? That should also fix this as that's handled by Floating UI

I've been trying to use the Popover component but it looks like it needs a trigger object (like a button) for it to work properly which doesn't exist because it is triggered by the right click.

I found a hacky way to get around it but the context menu still clips (because i'm still setting the position "manually").

```jsx
    <Popover isOpen={show} placement="top">
      <Popover.Trigger class="absolute left-0 top-0 bg-accent" />
      <Popover.Overlay />
      <Popover.Content>
        <div
          class={"absolute z-30 rounded-lg bg-thick-material"}
          style={{ top: pos()[1] + "px", left: pos()[0] + "px" }}
          ref={menu}
        >
          <div class="flex flex-col gap-1 rounded-lg bg-thick-material p-2">
            <For each={props.children}>{(child) => child}</For>
          </div>
        </div>
      </Popover.Content>
    </Popover>
```

Also, the Dropdown component in the settings uses Popover but also clips when the window is too small. Looking at the floating-ui docs it seems like we need to add [`flip()`](https://floating-ui.com/docs/flip) to the middleware in `Popover.tsx` and [`autoUpdate()`](https://floating-ui.com/docs/autoUpdate) so it updates when resizing the window.

I'll try looking into using floating-ui directly instead of the Popover component.

D0m1nos commented:
After messing around with floating-ui for a bit i found a solution.

First i tried using `flip()` to flip the context menu when it gets clipped, the problem is that apparently if the coordinates set in the middleware differ too much from the initial ones the `flip()` middleware doesn't work, it does the calculations but doesn't flip the menu. I tried calculating the new position manually but that caused other issues so instead i used `shift()`.

This clamps the menu position to the side instead of clipping.

![image](https://github.com/user-attachments/assets/07d522e0-3ba1-46df-849a-70c182233e14)

It's possible to flip the menu if we want to but it requires a couple of workarounds that could cause more issues and isn't worth it in my opinion.

duduBTW commented:
We should use the popover (and maybe also the `List`) component to avoid difference in styles on the project.

> I've been trying to use the Popover component but it looks like it needs a trigger object (like a button) for it to work properly which doesn't exist because it is triggered by the right click.

The idea is having a "more" button on the right of the cards that will show up when hover it:

`SongItem`:
<img width="466" alt="image" src="https://github.com/user-attachments/assets/9cc44739-fcac-42a0-be1d-ebaf911c49b7">

`PlaylistItem`:
<img width="543" alt="image" src="https://github.com/user-attachments/assets/ae4b1ffc-dff9-4691-8ea7-29f545dccd58">
_we don't have this component yet, but it has the same idea_

Right clicking will open the popover that is connect to the "more" trigger on the right.

You can add the necessary middlewares to floating-ui on the`Popover` component so everything works, just make sure it is customizable and it doesn't break the dropdown.

D0m1nos commented:
After some tinkering i managed to use the Popover component and should now work properly (at least on linux).

Also i wanted to apologize for not using the correct components/libraries initially (thanks duduBTW, brw and Tnixc for pointing me in the right direction), it's my first time working on a codebase this big so this is all new to me.

brw commented:
No need to apologize really! Your efforts are very much appreciated. We *want* this to be an environment for people to learn new things. There are no stakes, expectations or deadlines here, so don't worry about it honestly

duduBTW commented on src/renderer/src/components/song/context-menu/SongContextMenuItem.tsx:
```diff
@@ -18,12 +19,13 @@ const SongContextMenuItem: Component<SongContextMenuItemProps> = (props) => {
 
   return (
-    <button
+    <Button
```
This should not be a `Button` component, replace it with a regular "button".

The element is also way too big now.

<img width="747" alt="image" src="https://github.com/user-attachments/assets/161e2d38-1aa5-4520-a4fd-81ec360e06c8">

Meke it look like the other lists:

<img width="273" alt="image" src="https://github.com/user-attachments/assets/21695e9c-4a0c-4e0a-a947-1f9d6e0cee64">

duduBTW replied:
Hopefully we will be able to use the `<List />` here once that component is more flashed out, but just replacing the `<Button/>` and fixing the styles should be goo enough  for now.

duduBTW commented on src/renderer/src/components/song/song-item/SongItem.tsx:
```diff
@@ -57,32 +51,59 @@ const SongItem: Component<SongItemProps> = ({
-        </h3>
-        <p class="text-base text-subtext">{song.artist}</p>
+          <Popover.Trigger class="opacity-0 transition-opacity group-hover:opacity-100">
+            <Button variant={"ghost"} size={"icon"} class="z-50 mr-1 rounded-lg">
```
The `Popover.Trigger` is already a button, no need to put another button inside, replace it with something like:

``` diff
-  <Button variant={"ghost"} size={"icon"} class="z-50 mr-1 rounded-lg">
+ <div class="mr-2 grid aspect-square size-9 place-items-center rounded border-solid border-stroke bg-transparent p-1 text-text hover:bg-surface">
```

duduBTW commented on src/renderer/src/components/song/song-item/SongItem.tsx:
```diff
@@ -57,32 +51,59 @@ const SongItem: Component<SongItemProps> = ({
+            setLocalShow(false);
+          }}
+        >
+          {...children}
```
I don't think this the dropdown content should be retrieved from the children, could you move it to another prop instead, something like:

``` tsx
<SongItem
ContextMenu={<SongContextMenu>
                  <Show when={isQueueExist() === true}>
                    <PlayNext path={s.path} />
                  </Show>
                  <AddToPlaylist path={s.path} />
                </SongContextMenu>}
 />
```

D0m1nos commented:
All the issues mentioned should be fixed now. I also fixed the context menu clipping again and added a way to style the context menu items.

Thanks for the encouraging words brw, they really mean a lot to me :)

brw commented:
Some things I'm noticing at first glance:
- I don't think the context menu needs that extra 15px horizontal offset, it feels too far away imo. But we can always tweak it later
- I'd expect the context menu to open at the position of the cursor on right click (maybe a little difficult though)
- The trigger button should not disappear while the context menu is open
- I don't think you want to pass middlewares to the Popover component. Popover should add middlewares based on its props. You're setting the offset twice now for example: once through the offset prop, which adds an offset middleware, and once through another offset middleware.

The styling could use some work too but I suppose we might bring that in line with the actual design later.

brw commented:
Oh yea I also don't really like dynamically hiding/showing context menu items, e.g. the Play Next button only showing up when there's a queue. It should probably be disabled instead of hidden

D0m1nos commented:
> I don't think you want to pass middlewares to the Popover component.

You're right, that's also what the [docs](https://floating-ui.com/docs/computeposition#conditional) say but... (see next point)

> I'd expect the context menu to open at the position of the cursor on right click (maybe a little difficult though)

That's what i originally did by adding a custom middleware, i then removed it because dudu said (correct me if i misunderstood) "Right clicking will open the popover that is connect to the "more" trigger on the right."

If the context menu needs to be at the position of the cursor then i need some help implementing it, because i don't know how to do it without passing middlewares directly. Should i add a prop for a custom middleware? Should i add the middleware in the Popover component and a boolean prop to enable it?

Other than that i will implement the style changes mentioned.

brw commented:
> Should i add the middleware in the Popover component and a boolean prop to enable it

Yea that's what I meant.

Clicking on the trigger should open the menu belonging to the component at the trigger position.
Right clicking on a component should open the menu belonging to that the component at the cursor position.
Unless @duduBTW thinks differently about it but this seems pretty standard to me.

D0m1nos commented:
The context menu now opens on the mouse position when right-clicking and the unclickable items are now disabled instead of hidden.

Quick question about the trigger button, should it only stay visible when left-clicking it and stay hidden when right-clicking elsewhere? Right now it looks like this when right-clicking near it:

![image](https://github.com/user-attachments/assets/515bf6e6-8515-40c5-bb38-1a124de8b691)

brw commented:
Latest changes seem wrong to me. The trigger now always disappears as soon as the menu is opened either through right click or through the trigger.

I think the problem is just with focus. The trigger should always be shown as long as you're hovered over the song, even if the menu is open. And if you want to follow what Spotify does it should also stay shown when moving the cursor into the menu if the menu was opened by clicking on the trigger.

D0m1nos commented:
That's weird, it was not what i intended to do and when i tried it the trigger didn't disappear when left clicking.

Anyways, i reverted the change for now, this is only a minor visual concern that i think should be talked about when we will implement the newer UI.

When the Popover appears it traps all of the window focus in the Popover.Overlay, so if we wanted to mimic Spotify we would have to remove that but then it would cause other issues like:

![image](https://github.com/user-attachments/assets/13c82226-d50f-4da2-8220-7a00b4a68406)

brw commented:
There is probably a good solution for that but yea I guess something to look at later

brw commented on src/renderer/src/scenes/main-scene/MainScene.tsx:
```diff
@@ -173,20 +173,26 @@ const QueueModal: Component = () => {
+      if (
+        queueModal &&
+        !queueModal.contains(event.target as Node) &&
+        event.target.closest(".popover-overlay") === null
```
Not too sure about this but I guess it'll do for now. Just have to rework all of this logic at some point I think

D0m1nos replied:
This was to solve a bug where clicking on the context menu (opened on a song in the queue) would close both the context menu and the queue without doing the clicked action. The main scene is being completely remade in #88 so this is a temporary fix.

Tnixc commented:
Could we make it so that if you right click on another song while a menu is already open dismiss the currently open the menu and open the menu for that song?
(basically open menu while another menu is open)

duduBTW commented:
> Could we make it so that if you right click on another song while a menu is already open dismiss the currently open the menu and open the menu for that song? (basically open menu while another menu is open)

I don't think implementing that effort right now, I also like keeping the interactions on the popover consistent (clicking away closes it).

brw commented:
Should definitely do that at some point though. Definitely how people would expect it to work coming from other apps
